### PR TITLE
Refactor AuthenticationResponse.errorResponse method

### LINKED ISSUE #9: Refactor AuthenticationResponse.errorResponse method
The static `AuthenticationResponse.errorResponse` method, as part of the `validateResponse` flow, is meant to convert the standard [OAuth2 error parameters](https://tools.ietf.org/html/rfc6749#page-45) (`error` code, `error_description` and `error_uri`) into an actual Error object and to throw it. 

(The AuthenticationResponse.validateResponse flow is also supposed to handle IDP errors returned to the callback endpoint as the OIDC/OAuth2 error params.)

Action items:
* Move `errorResponse` up earlier in the control flow chain -- currently it's called after `matchRequest` and `validateStateParam`, but it should be called first thing.
* Currently it only uses the OAuth2 `error` param, so need to add the `error_description` and `error_uri` params.
* Consider using [HttpError](https://github.com/moll/js-standard-http-error) or something OAuth2-specific.

### PR DESCRIPTION
**Pull request by dmitrizagidulin**
Refactor `AuthenticationResponse.errorResponse` method (see issue #9 for details).
Closes #9.

### CODE REVIEW

**Comment by dmitrizagidulin** [comment]
This is a minor refactoring that's adding comments and setting up better error reporting down the line.

cc @kjetilk / @RubenVerborgh for review.

File: src/AuthenticationResponse.js [hunk 1]
```diff
@@ -82,6 +102,37 @@ class AuthenticationResponse {
...
+   *
+   * @todo Figure out HTTP status code (typically 400, 401 or 403)
+   *   based on the OAuth2/OIDC `error` code, probably using an external library
+   */
+  static errorResponse (response) {
+    const errorCode = response.params.error
+
+    if (errorCode) {
+      const errorParams = {}
+      errorParams['error'] = errorCode
```
Comment (RubenVerborgh) [inline src/AuthenticationResponse.js]: Why the brackets?

File: src/AuthenticationResponse.js [hunk 2]
```diff
@@ -82,6 +102,37 @@ class AuthenticationResponse {
...
+   * @throws {Error} If response params include the OAuth2 'error' param,
+   *   throws an error based on it.
+   *
+   * @returns {AuthenticationResponse} Chainable
+   *
+   * @todo Figure out HTTP status code (typically 400, 401 or 403)
+   *   based on the OAuth2/OIDC `error` code, probably using an external library
+   */
+  static errorResponse (response) {
+    const errorCode = response.params.error
```
Comment (RubenVerborgh) [inline src/AuthenticationResponse.js]: Passing things around this way creates global state.

File: src/AuthenticationResponse.js (hunk 1 above)
Comment (RubenVerborgh) [reply src/AuthenticationResponse.js]: Also, weird property name. error.error?

File: src/AuthenticationResponse.js [hunk 3]
```diff
@@ -321,14 +356,25 @@ class AuthenticationResponse {
...
+   * @param response.params {object}
+   * @param [response.params.id_token] {string} IDToken encoded as a JWT
+   *
+   * @returns {AuthenticationResponse} Chainable
    */
   static decodeIDToken (response) {
     let jwt = response.params.id_token
 
-    if (jwt) {
+    try {
```
Comment (RubenVerborgh) [inline src/AuthenticationResponse.js]: Do we want to get rid of the check? Missing jwt seems to warrant another error.

File: src/AuthenticationResponse.js [hunk 4]
```diff
@@ -82,6 +102,37 @@ class AuthenticationResponse {
...
+   */
+  static errorResponse (response) {
+    const errorCode = response.params.error
+
+    if (errorCode) {
+      const errorParams = {}
+      errorParams['error'] = errorCode
+      errorParams['error_description'] = response.params['error_description']
+      errorParams['error_uri'] = response.params['error_uri']
+      errorParams['state'] = response.params['state']
```
Comment (RubenVerborgh) [inline src/AuthenticationResponse.js]: Also weird names, with underscore. What is state?

Comment (dmitrizagidulin) [reply src/AuthenticationResponse.js]: Heh, yes, they are weird names, but they're straight from the [OAuth2 Error Response](https://tools.ietf.org/html/rfc6749#section-5.2) spec. And `state` is from the OIDC spec.

File: src/AuthenticationResponse.js (hunk 2 above)
Comment (dmitrizagidulin) [reply src/AuthenticationResponse.js]: I don't think so, in this particular case. `response` isn't the Express `ServerResponse` object. it's a `AuthenticationResponse` instance, and the params are parsed into it.

File: src/AuthenticationResponse.js (hunk 3 above)
Comment (dmitrizagidulin) [reply src/AuthenticationResponse.js]: I think we do. If `id_token` is missing, the flow gets short-circuited earlier in `validateIDToken()`. So this method will not be called.

Even for standalone reasons, if it's missing, that'll be caught in the `try` / `catch`

**Comment by dmitrizagidulin** [comment]
@RubenVerborgh we good, re the comments?

**Pull request merged**